pkg/cdi/cache.go

```diff
@@ -340,7 +340,7 @@ func (c *Cache) GetDevice(device string) *Device {
 	c.Lock()
 	defer c.Unlock()
 
-	c.refreshIfRequired(false)
+	_, _ = c.refreshIfRequired(false) // we record but ignore errors
```

Comment:
Are there places where we actually check the error return values for the internal calls?

Reply:
> Can we call `c.Refresh(`) here?

No, I don't think so. `Refresh()` is a locked `refreshIfRequired()` which always returns collected errors if it ended up doing a refresh. So it would be the same if it did not deadlock itself.

Reply:
> Are there places where we actually check the error return values for the internal calls?

We check and propagate errors back on the caller only on the externally triggered calls (`Refresh()`), not on the internally triggered ones. But we handle both with a single code path.

Reply:
Was just able to look through the code. Thanks for confirming.

Reply:
> @klihub
> 
> > Refresh() is a locked refreshIfRequired() which always returns collected errors if it ended up doing a refresh. So it would be the same but would deadlock itself.
> 
> Sorry for not being clear. I actually meant using `Refresh` and remove locking on lines 340 and 341. Are those locks needed for anything except locking `refreshIfRequired`?

The locks are for providing atomicity between refresh scanning Spec directories and updates the Specs/the Cache wrt. any other externally triggered access to the cache accessing the same data. IOW, you either get access to the data in the cache before a refresh cycle or after, but not while a refresh is happening.